### PR TITLE
Add `pswf_operator` to dspec

### PR DESCRIPTION
**Pull request by tyler-a-cox**
This PR adds another operator, `pswf_operator`, to `hera_filters`. The `pswf_operator` is a continuous version of the `dpss_operator` that allows the user to sample points off of uniform a grid while still maintaining the useful properties of the `dpss_operator`. Particularly useful for `nucal`. Implementation from the following [paper](https://reader.elsevier.com/reader/sd/pii/S106352030400017X?token=5389AFAD9862A5E3505E405E1BEC72C9AE0CE3638120545A122BC35A5CAC9CF57B777B2233FD876DF208F43B9C7AC113&originRegion=us-east-1&originCreation=20220907192434).

### CODE REVIEW

File: hera_filters/dspec.py [hunk 1]
```diff
@@ -2120,6 +2120,215 @@ def fit_solution_matrix(weights, design_matrix, cache=None, hash_decimal=10, fit
...
+
+            # Apply filter center to design matrix
+            amat.append(
+                pswf_vectors
+                * np.exp(2j * np.pi * xg[:, None] * fc / (2 * np.pi * (xmax - xmin)))
+            )
+
+        # Stack design matrices
+        if len(amat) > 1:
+            amat = np.hstack(amat)
```
Comment (aewallwi) [inline hera_filters/dspec.py:2325]: This needs coverage.

**Review by aewallwi — CHANGES_REQUESTED** [review]
Looks good mostly. Main items are links to literature explaining how we calculate pswf from legendre polynomials and also the eigenvals. A few missing coverage lines too.

File: hera_filters/dspec.py [hunk 2]
```diff
@@ -2120,6 +2120,215 @@ def fit_solution_matrix(weights, design_matrix, cache=None, hash_decimal=10, fit
                 cache[opkey] = None
     return cache[opkey]
 
+def _calculate_amat(k, c):
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_filters/dspec.py:2123]: Added a description of this function

File: hera_filters/dspec.py [hunk 3]
```diff
@@ -2120,6 +2120,215 @@ def fit_solution_matrix(weights, design_matrix, cache=None, hash_decimal=10, fit
...
+    amat[ks + 2, ks] = (
+        (ks + 2) * (ks + 1) / ((2 * ks + 3) * np.sqrt((2 * ks + 1) * (2 * ks + 5))) * c ** 2
+    )
+    eigenvals, eigenvecs = np.linalg.eigh(amat[:k, :k])
+    return eigenvals, eigenvecs
+
+def _normalized_legendre(x, kmax):
+    """
+    Fast computation of the first kmax normalized Legendre polynomials evaluated at points
+    x using the recurrence relation.
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_filters/dspec.py:2157]: Added a link to the recurrence relation

File: hera_filters/dspec.py [hunk 4]
```diff
@@ -2120,6 +2120,215 @@ def fit_solution_matrix(weights, design_matrix, cache=None, hash_decimal=10, fit
...
+    do not need to be equally spaced (unlike DPSS operator). Users can specify how the
+    PSWF series fits are cutoff in each delay-filtering window with one (and only one)
+    of two conditions: eigenvalues in sinc matrix fall below a thresshold (eigenval_cutoff) or
+    user specified number of PSWF terms (nterms).
+
+    Parameters
+    ----------
+    x: array-like
+        x values to evaluate operator at
+    filter_centers: array-like
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_filters/dspec.py:2201]: Good point. I'll change that docstring

File: hera_filters/dspec.py [hunk 5]
```diff
@@ -2120,6 +2120,215 @@ def fit_solution_matrix(weights, design_matrix, cache=None, hash_decimal=10, fit
...
+            p2 = polynomials[n - 2]
+            pn = ((2 * n - 1) * x * p1 - (n - 1) * p2) / float(n)
+            polynomials.append(pn)
+
+    polynomials = np.sqrt(np.arange(kmax) + 0.5) * np.transpose(polynomials)
+    return polynomials
+
+def pswf_operator(
+    x, filter_centers, filter_half_widths, eigenval_cutoff=None, nterms=None,
+    cache=None, xmin=None, xmax=None, hash_decimal=10,
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_filters/dspec.py:2188]: Added a link

File: hera_filters/dspec.py [hunk 6]
```diff
@@ -2120,6 +2120,215 @@ def fit_solution_matrix(weights, design_matrix, cache=None, hash_decimal=10, fit
...
+    filter_centers: array-like
+        list of floats of centers of delay filter windows in nanosec
+    filter_half_widths: array-like
+        list of floats of half-widths of delay filter windows in nanosec
+    cache: dictionary, optional
+        dictionary for storing operator matrices with keys
+        tuple(x) + tuple(filter_centers) + tuple(filter_half_widths)\
+         + (series_cutoff_name,) = tuple(series_cutoff_values)
+    eigenval_cutoff: list of floats, optional
+        list of sinc matrix eigenvalue cutoffs to use for included dpss modes.
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_filters/dspec.py:2210]: Changed these to PSWF

File: hera_filters/dspec.py [hunk 7]
```diff
@@ -2120,6 +2120,215 @@ def fit_solution_matrix(weights, design_matrix, cache=None, hash_decimal=10, fit
...
+    hash_decimal: int
+        number of decimals to round for floating point dict keys.
+
+
+    Returns
+    ----------
+    2-tuple
+    First element:
+        Design matrix for DPSS fitting.   Ndata x (Nfilter_window * nterm)
+        transforming from DPSS modes to data.
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_filters/dspec.py:2226]: Changed to pswf

File: hera_filters/dspec.py [hunk 8]
```diff
@@ -2120,6 +2120,215 @@ def fit_solution_matrix(weights, design_matrix, cache=None, hash_decimal=10, fit
...
+
+def pswf_operator(
+    x, filter_centers, filter_half_widths, eigenval_cutoff=None, nterms=None,
+    cache=None, xmin=None, xmax=None, hash_decimal=10,
+):
+    """
+    Calculates PSWF operator with multiple delay windows to fit data. Frequencies
+    do not need to be equally spaced (unlike DPSS operator). Users can specify how the
+    PSWF series fits are cutoff in each delay-filtering window with one (and only one)
+    of two conditions: eigenvalues in sinc matrix fall below a thresshold (eigenval_cutoff) or
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_filters/dspec.py:2194]: The algorithm isn't explicitly using a sinc matrix, but the eigenvalues are still of a sinc function, so I don't think it's totally incorrect to say.

Screen Shot 2022-09-23 at 7 16 59 AM

File: hera_filters/dspec.py (hunk 1 above)
Comment (tyler-a-cox) [reply hera_filters/dspec.py:2338]: All lines for the new functions should be covered now

File: hera_filters/dspec.py [hunk 9]
```diff
@@ -2120,6 +2120,215 @@ def fit_solution_matrix(weights, design_matrix, cache=None, hash_decimal=10, fit
...
+        cache = {}
+
+    # Conditions for halting
+    crit_labels = ['eigenval_cutoff', 'nterms']
+    crit_list = [eigenval_cutoff, nterms]
+    crit_provided = np.asarray([not crit is None for crit in crit_list]).astype(bool)
+    crit_provided_name = [label for m,label in enumerate(crit_labels) if crit_provided[m]]
+    crit_provided_value = [crit for m,crit in enumerate(crit_list) if crit_provided[m]]
+    if np.count_nonzero(crit_provided) != 1:
+        raise ValueError(
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_filters/dspec.py:2254]: Also covered this. I copy over the line from the dpss_operator function and didn't change it to pswf_operator. I can remove those other modes from DPSS in another PR.

**Pull request merged**